### DISCUSSION TITLE
Utilizing shortLinking

### DISCUSSION THREAD

**Discussion by workmap** [TerriaJS/terriajs#6580] [Q&A]
Greetings I am trying to implement short linking in terriamap share. I see that there is a `shareURLPrefixes` property in `devserverconfig.json` ( [terriajs-server shareURLPrefixes configuration](https://github.com/TerriaJS/terriajs-server/blob/master/serverconfig.json.example#L139-L155) ). I used the aws s3 configuration but nothing happens. I gave IAM permissions to my terriamap container to handle the s3 gut and get requests. I have verified that `shareUrlPrefix` is loaded into `terriajs-server`. However the url is not shortened and nothing is written to s3. 

Has anyone worked with shareUrlPrefixes for s3?

**Comment by nf-s** [comment]
Hi @workmap 

Are you able to share your configuration (without secrets of course)?

The following config is working for me on latest version of terriajs

```json
{
  "newShareUrlPrefix": "s",
  "shareUrlPrefixes": {
    "s": {
      "service": "s3",
      "region": "ap-southeast-2",
      "bucket": "xxx",
      "accessKeyId": "xxx",
      "secretAccessKey": "xxx"
    }
  }
}
```

**Reply by workmap** [reply]
Hi @nf-s, 
Thanks for the reply. Here is my configuration:

```
"shareUrlPrefixes": {
        // A blank prefix resolves without the separator: .../#share=12345
        "s": {
            "service": "s3",
            // Required: the AWS region to access
            "region": "us-east-1",

            // Required: an existing S3 bucket in which to store objects
            "bucket": "gp-sit-us-east-1-terriamap-sharing"

            // Optional: The access key ID and secret access key of a user with S3 getObject and putObject permission on the above bucket
            // If not provided here, you must ensure they're available as environment variables or in a shared credentials file.
            // See http://docs.aws.amazon.com/AWSJavaScriptSDK/guide/node-configuring.html
            
            // "accessKeyId": "",
            // "secretAccessKey": "",

            // Optional: The length of the random share key to generate (not including prefix), up to 54 characters. Defaults to the full length.
            // "keyLength": 16
        }
    }
```

I am not providing the `accessKeyId` or the `secretAccessKey` but instead providing IAM permissions to the container. Can this be done or do I need to use the `accessKeyId` and the `secretAccessKey`.  

I have also tried with my `accessKeyId` and `secretAccessKey` but the results are the same. 

When working there should be a short URL here right? 
<img width="453" alt="Screen Shot 2022-10-04 at 1 23 40 PM" src="https://user-images.githubusercontent.com/61807178/193885886-d8c0e7ec-7359-4d62-b9c8-a9a92d529366.png">

**Reply by nf-s** [reply]
Providing IAM permissions to VM/container should be fine.

Yes the short URL should appear there. 

Are you getting any error messages from the POST request made to terriajs-server (on `/share`)

Also, do you have `"newShareUrlPrefix": "s",` in your configuration?

**Reply by workmap** [reply]
Thanks for the help! Ahhh the newShareUrlPrefix was definitely one of the problems.  Now I'm able to use the share feature with accessKeyId and secretAccessKey. However when I try to use IAM permissions instead I get the following error:

`Missing credentials in config, if using AWS_CONFIG_FILE, set AWS_SDK_LOAD_CONFIG=1`

If accessKeyId and secretAccessKey are not provided is AWS_CONFIG_FILE required? 

I could not locate the POST request to terriajs-server. Is that being made from terriajs or Terriamap? 

We are currently using this commit `ad0689e82f10dfb880e37d440c0ac668e1090dd6` of Terriamap.
